heroku/data_source_heroku_app.go

```diff
@@ -0,0 +1,106 @@
+package heroku
+
+import (
+	"context"
+	"github.com/cyberdelia/heroku-go/v3"
+	"github.com/hashicorp/terraform/helper/schema"
+	"time"
+)
+
+func dataSourceHerokuApp() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceHerokuAppRead,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"space": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Default:  nil,
+			},
+
+			"region": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Default:  nil,
+			},
+
+			"stack": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Default: nil,
+			},
+
+			"buildpacks": {
+				Type: schema.TypeList,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Computed: true,
+			},
+
+			"config_vars": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeMap,
+				},
+			},
+
+			"git_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"web_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"organization": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Default:  nil,
+			},
+		},
+	}
+}
+
+func dataSourceHerokuAppRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(*heroku.Service)
+
+	name := d.Get("name").(string)
+	app, err := client.AppInfo(context.TODO(), name)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(time.Now().UTC().String())
+
+	// Flag organization apps by setting the organization name
+	if app.Organization != nil {
+		d.Set("organization", []map[string]interface{}{
+			{"name": app.Organization.Name},
+		})
+	} else {
+		d.Set("organization", "")
```

Comment:
Should this be an empty map, or perhaps `nil` when not present, or does a space work?

Reply:
@jmervine so far in my testing - there are problems with using `nil`. Basically, I've got this terraform script:

```
data "heroku_app" "parent" {
  name = "${var.parent_app}"
}

resource "heroku_app" "child" {
  name = "${var.child_app}"
  region = "us-east"

  config_vars {
    ORGANIZATION = "${data.heroku_app.parent.organization}"
    REGION = "${data.heroku_app.parent.region}"
    SPACE = "${data.heroku_app.parent.space}"
    STACK = "${data.heroku_app.parent.stack}"
    GIT_URL = "${data.heroku_app.parent.git_url}"
    WEB_URL = "${data.heroku_app.parent.web_url}"
  }
}
```

I finding that if `organization` (or any other attribute) is nil, the terraform script will fail to apply, with errors like this:

```
* heroku_app.child: Resource 'data.heroku_app.parent' does not have attribute 'space' for variable 'data.heroku_app.parent.space'
```

So the question here is: Should the script actually fail if the app doesn't *have an organization*? My belief is "no".  I think it's better to return an empty string.

I see three possiblities:
1. I'm doing something wrong and it *is* possible to have a nil value for an attribute but still have attribute work if referenced (so far, though, I don't think I'm wrong about this).
2. You actually shouldn't try to reference an attribute that is nil (this seems silly though - the script becomes very brittle if I want to apply the same recipe to different resources where some of them have an `organization` but others don't).
3. What I'm doing in the script is correct, as is the way I'm continuing to expose "nil" attributes as empty strings.

BTW it looks like "how to represent nil attributes in terraform" is 
 a problem others have encountered: https://github.com/hashicorp/terraform/issues/5471#issuecomment-350115135

Reply:
I tend to agree. Origination is optional and should be treated as such.

Reply:
So if I'm understanding you correctly, if you don't handle the `nil`, it errors. How does the heroku lib handle the empty string? My other suggestion was to set it to an empty map, e.g.
```golang
d.Set("organization", []map[string]interface{}{})
```
Might that be better?

Reply:
Err wait. What are you agreeing with? There were three possibilities, you can't agree with all of them. I'm not sure what you mean by "origination is optional and should be treated as such".

Based on the rest of your comment, I *think* you're agreeing with option 2.

Reply:
@jmervine it looks like, for consistency, option 2 is appropriate. Here's what happens if you try to reference "organization" for a heroku_app resource that doesn't have one:

```

* heroku_app.child: Resource 'heroku_app.managed' does not have attribute 'organization' for variable 'heroku_app.managed.organization'
```

... This is based on this terraform script:

```
data "heroku_app" "parent" {
  name = "${var.parent_app}"
}

resource "heroku_app" "managed" {
  name = "shea-managed-1"
  region = "us-east"
}

resource "heroku_app" "child" {
  name = "${var.child_app}"
  region = "us-east"

  config_vars {
    ORGANIZATION = "${data.heroku_app.parent.organization}"
    REGION = "${data.heroku_app.parent.region}"
    SPACE = "${data.heroku_app.parent.space}"
    STACK = "${data.heroku_app.parent.stack}"
    GIT_URL = "${data.heroku_app.parent.git_url}"
    WEB_URL = "${data.heroku_app.parent.web_url}"

    MANAGED_ORGANIZATION = "${heroku_app.managed.organization}"
  }
}
```

Reply:
So, I'm going to simply fail if users try to reference nil attributes. I think this is a shortcoming of terraform, but it's not one that is really possible to work around in any good way at this point.

Reply:
Is there even possible to pass a nil value via Terraform? I thought it was omit the config or set it to a string or map. Are empty maps (or blocks in terraform DSL) even possible?

I was agreeing with 'Should the script actually fail if the app doesn't have an organization? My belief is "no".'